[v10] Add label and legend `headingLevel` option

colinrotherham opened the pull request:
## Description

The Nunjucks option `isPageHeading` sets whether the label or legend also acts as the heading for the page.

This PR adds a new `headingLevel` Nunjucks option as a replacement for `isPageHeading` when a custom heading level is necessary, for example when a label or legend acts as the heading for a landmark:

```patch
<aside>
  {{ input({
    label: {
      text: 'Search',
-     isPageHeading: true
+     headingLevel: 2
    }
  }) }}

  <!-- // … -->
</aside>
```

This aligns with support already in NHS.UK React components

See related PR and issue over on GOV.UK Frontend

* https://github.com/alphagov/govuk-frontend/pull/5240
* https://github.com/alphagov/govuk-frontend/issues/6471

## Checklist

- [ ] Tested against our [testing policy](https://github.com/nhsuk/nhsuk-frontend/blob/master/docs/contributing/testing.md) (Resolution, Browser & Accessibility)
- [x] Follows our [coding standards and style guide](https://github.com/nhsuk/nhsuk-frontend/blob/master/docs/contributing/coding-standards.md)
- [x] CHANGELOG entry

colinrotherham committed:
Add legend `headingLevel` Nunjucks param

colinrotherham committed:
Add label `headingLevel` Nunjucks param

colinrotherham committed:
Add examples for label `headingLevel`

colinrotherham committed:
Add examples for legend `headingLevel`

frankieroberto commented on packages/nhsuk-frontend/src/nhsuk/components/label/template.njk:
```diff
@@ -1,18 +1,46 @@
+
+{%- if params.size and params.size in ["s", "m", "l", "xl"] and "nhsuk-label--" not in classNames %}
+  {% set classNames = classNames ~ " nhsuk-label--" ~ params.size %}
+{% endif %}
```
I think these should be in the opposite order, so that `classes` are appended last? 

(that gives them higher priority, so that if you specified `size: 'xl'` and `classes: 'nhsuk-label--l'` then it'd render as large)

I think we should also drop `"nhsuk-label--" not in classNames` as that’d false-positive if you had another modifier like `nhsuk-label--somethingelse`?

colinrotherham commented:
Sharing an update from the GOV.UK team

They're using custom heading levels in [smart answers](https://gds.blog.gov.uk/2012/02/16/smart-answers-are-smart/) and (internal) publishing applications

But other than that they're not widely used

>* checkboxes allows a legend to contain a H1 with a toggle option: https://components.publishing.service.gov.uk/component-guide/checkboxes
>* fieldset allows a custom size and heading level for the legend but I don't know if we're actually using it anywhere (it's not clearly documented): https://components.publishing.service.gov.uk/component-guide/fieldset#with_heading
>* file upload allows a custom size and heading level but this isn't on the public site, used by publishing applications: https://components.publishing.service.gov.uk/component-guide/file_upload#with_label_as_heading
>* input allows a custom size and heading level but again don't think this is actually used anywhere: https://components.publishing.service.gov.uk/component-guide/input#with_label_as_heading
>* label allows allows a H1 to wrap the label, but only as a toggle option: https://components.publishing.service.gov.uk/component-guide/label#as_page_heading
>* radios allows a custom size and heading level, I believe this is used in smart answers: https://components.publishing.service.gov.uk/component-guide/radio#with_custom_heading_size
>* textarea has a toggle to allow the label to be wrapped in a H1 but this is only used in publishing: https://components.publishing.service.gov.uk/component-guide/textarea#with_label_as_page_heading

colinrotherham committed:
Add changelog entry

colinrotherham commented:
@copilot rebase and resolve the merge conflicts on this branch. Only the top 5 commits should remain

frankieroberto reviewed (approved):
Looks good to me.

Could we perhaps keep `isPageHeading: true` (or `heading: true`) around a shortcut, which could perhaps set both `headingLevel: 1` and `size: "l"`?

colinrotherham commented:
Thanks @frankieroberto

Makes sense to me to rename `isPageHeading: true` to `heading: true`

Then it fits our naming convention for other boolean options

We can support both until a future major release

paulrobertlloyd commented:
There’s an API inconsistency here: in the case of labels and legends `heading` accepts a `boolean`, whereas in other components it accepts a `string` or `object`. We’ve also deprecated `headingLevel` on the card component.

I wonder if the solution is something more along the lines of this:

## Label (not a heading)

```njk
label: {
  text: "First name",
  size: "m"
}
```

```html
<label class="nhsuk-label nhsuk-label--m">First name</label>
```

## Label (simple heading)

```njk
label: {
  heading: "First name",
  size: "m"
}
```

```html
<h1 class="nhsuk-label-wrapper">
  <label class="nhsuk-label nhsuk-label--m">First name</label>
</h1>
```

- Any value for the `text` option is superseded by `heading`

## Label (complex heading)

```njk
label: {
  heading: {
    text: "First name",
    level: 2,
    size: "m"
  }
}
```

```html
<h2 class="nhsuk-label-wrapper">
  <label class="nhsuk-label nhsuk-label--m">First name</label>
</h2>
```

- Any value for the `text` option is superseded by `heading.text`
- Any value for the `size` option is superseded by `heading.size`

## `isPageHeading`

`isPageHeading` can then be implemented thus:

```diff
  label: {
-   isPageHeading: true,
-   text: "First name",
+   heading: "First name"
  }
```

```html
<h1 class="nhsuk-label-wrapper">
  <label class="nhsuk-label nhsuk-label--l">First name</label>
</h1>
```

colinrotherham commented:
@paulrobertlloyd Interesting points!

We'd be doubling up on a few things though?

* Setting `caption` at the top level (versus `heading.caption`)
* Setting `size` at the top level (versus `heading.size`)

E.g. A label can have a caption without it being a heading:

```njk
{{ label({
  caption: "About you",
  text: "What is your full name?",
  size: "l"
}) }}
```

Don't think we can put all the options at both the top level and within `heading` can we?

```njk
{{ label({
  caption: "About you",
  text: "What is your full name?",
  size: "l",
  heading: {
    caption: "About me",
    text: "What is my full name?",
    size: "l",
    level: 1
  },
  headingLevel: 2
}) }}
```

I'm viewing `heading: true` a bit like `card: true` in the table component

Yes it's a boolean, but you can also pass in all the card options too

We can update the `macro-options.mjs` to show it's a heading ~boolean~ object?

colinrotherham commented:
@anandamaryon1 Any thoughts on @paulrobertlloyd's simple heading proposal?

A little bit like sortable tables, are we happy with `text` → `heading` being the default opt-in for headings?

```patch
  {{ label({
-   text: "What is your full name?",
-   isPageHeading: true,
+   heading: "What is your full name?",
+   headingLevel: 2,
    size: "l"
  }) }}
```

Yet also allowing boolean `heading: true`

```patch
  {{ label({
    text: "What is your full name?",
-   isPageHeading: true,
+   heading: true,
+   headingLevel: 2,
    size: "l"
  }) }}
```

Then for the power users, allowing `heading` objects for compatibility?

```patch
  {{ label({
-   text: "What is your full name?",
-   isPageHeading: true,
+   heading: {
+     text: "What is your full name?"
+     level: 2,
+     size: "l"
+   }
-   size: "l"
  }) }}
```

If so, this is what I've pushed up

anandamaryon1 commented:
Looks good to me. I'd worried a little bit before about it being confusing to have multiple ways to do the same thing, but probably most will just copy whatever we do in the design system examples, and those inclined to dig into the options will appreciate the flexibility. 

Happy with this as-is so long as we've covered all the combinations, eg. `heading.text` matched with `label.size` etc. (but it seems you're on it).

frankieroberto commented:
Looks good to me!

paulrobertlloyd commented:
Is it wise to have both `headingLevel` and `heading.level`? Should it be that once you want to add more heading options, you need to use the nested `heading` object?

And how have we ended up with `heading.headingLevel`?

paulrobertlloyd commented:
Also, is there a particular reason why reference images have changed, or does that relate to another change in this PR?

colinrotherham commented:
> Is it wise to have both `headingLevel` and `heading.level`? Should it be that once you want to add more heading options, you need to use the nested `heading` object?

@paulrobertlloyd Ananda wanted `title.headingLevel` (not `title.level`) for the notification banner title (in the title bar) but we already shipped that in v10.6.0

We agreed to alias it for compatibility and it made reasonable sense to alias it in other places

Don't have to. We can limit that?

colinrotherham commented:
> Also, is there a particular reason why reference images have changed, or does that relate to another change in this PR?

@paulrobertlloyd Think it's just new caption examples, like in the fieldset (legend caption) fixtures

paulrobertlloyd commented:
> it's just new caption examples, like in the fieldset (legend caption) fixtures

🤦🏼‍♂️

paulrobertlloyd commented:
> We agreed to alias it for compatibility and it made reasonable sense to alias it in other places

I know we’re already juggling a lot of aliases, deprecations etc., not sure having more helps in the long run. Is `title.level` really _that_ bad?

colinrotherham commented:
>I know we’re already juggling a lot of aliases, deprecations etc., not sure having more helps in the long run. Is `title.level` really that bad?

@paulrobertlloyd Didn't necessarily think so at the time

But I can make `headingLevel` only available for `label`, `legend` and the notification banner `title`?

Then we're good to go?

paulrobertlloyd commented:
Without being a _complete_ pain (!) might it be possible to see a full list of all heading related options (including deprecated and those proposed in this PR)?

colinrotherham commented:
@paulrobertlloyd Not a pain at all

Heading related changes in this PR for v10.6.1

### Heading component
* Added `headingLevel` integer to alias `level`

### Label component
* Added `heading` string or heading component with nested:
  * `text` string
  * `html` string
  * `visuallyHiddenText` string
  * `caption` string or caption component
  * `size` string
  * `level` integer

* Added `headingLevel` integer to alias `heading.level`
* Added `caption` string or caption component to alias `heading.caption`
* Added `visuallyHiddenText` string
* Deprecated `isPageHeading` boolean

### Legend component
* Added `id` string
* Added `heading` string or heading component with nested:
  * `text` string
  * `html` string
  * `visuallyHiddenText` string
  * `caption` string or caption component
  * `size` string
  * `level` integer

* Added `headingLevel` integer to alias `heading.level`
* Added `caption` string or caption component to alias `heading.caption`
* Added `visuallyHiddenText` string
* Added `attributes` object
* Deprecated `isPageHeading` boolean

**Note:** Legend `caption.placement` is limited to `"start"`, `"end"` or `"after".` We check for `"before"` and replace it with `"start"` due to the HTML restriction to have `<legend>` as the first child of `<fieldset>`

### Table component
* Added `caption.caption` string or caption component (i.e. table caption with heading caption)

### Warning callout
* Added `headingClasses` string documentation, missed from v10.3.0, deprecated v10.6.0

colinrotherham commented:
☝️ Although that was just for this PR

Think this is everything, heading related changes from v10.6.0

### Card component
* Added `heading` string or heading component
* Deprecated `headingHtml` string
* Deprecated `headingClasses` string
* Deprecated `headingSize` string
* Deprecated `headingLevel` integer
* Deprecated `headingId` string
* Deprecated `headingVisuallyHiddenText` string

### Do and don't list component
* Added `title` string or heading component
* Deprecated `title` string
* Deprecated `headingLevel` integer

### Error summary component
* Added `heading` string or heading component
* Deprecated `titleText` string
* Deprecated `titleHtml` string

### Footer component
* Added `navigation.heading` string or heading component
* Deprecated `navigation.title` string
* Added `meta.visuallyHiddenText` string
* Deprecated `meta.visuallyHiddenTitle` string

### Hero component
* Added `heading` string or heading component
* Deprecated `headingClasses` string
* Deprecated `headingSize` string
* Deprecated `headingLevel` integer

### Notification banner component
* Added `title` string or heading component
* Deprecated `titleId` string
* Deprecated `titleText` string
* Deprecated `titleHtml` string
* Deprecated `titleHeadingLevel` string
* Added `heading` string or heading component

### Panel component
* Added `heading` string or heading component
* Deprecated `titleText` string
* Deprecated `titleHtml` string
* Deprecated `titleSize` string
* Deprecated `titleClasses` string
* Deprecated `headingLevel` integer

### Tabs component
* Added `caption` string or caption component
* Deprecated `captionClasses` string
* Deprecated `captionSize` string

### Tabs component
* Added `visuallyHiddenText` string
* Deprecated `title` string

### Task list component
* Added `heading` string or heading component
* Deprecated `title` heading component

### Warning callout component
* Added `heading` string or heading component
* Deprecated `headingLevel` integer
* Deprecated `title` heading component

paulrobertlloyd commented:
I’m for not having `label.headingLevel`. It should be a consistent API policy that “`heading` (string) is a large sized level 1 heading. If you want to change those defaults, use `heading` (object) instead”. So to change a label’s heading level you would use `label.heading.level`.

colinrotherham commented:
@paulrobertlloyd I've removed `headingLevel` from labels and legends

But as requested by @anandamaryon1 I've kept it for the `title.headingLevel` → `title.level` alias

Which means this PR now has our default upgrade path as `text` to `heading` strings:

```patch
  {{ label({
-   text: "What is your full name?",
+   heading: "What is your full name?",
-   isPageHeading: true,
    size: "l"
  }) }}
```

Yet we'd also allow boolean `heading: true` to preserve existing `text` or `html` options:

```patch
  {{ label({
    text: "What is your full name?",
-   isPageHeading: true,
+   heading: true,
    size: "l"
  }) }}
```

Then for the power users you can either partially add `heading` nested options:

```patch
  {{ label({
    text: "What is your full name?",
-   isPageHeading: true,
+   heading: {
+     id: "custom-id"
+     level: 2
+   },
    size: "l"
  }) }}
```

Or fully embrace them, allowing the same `heading` options to be shared with other components:

```patch
  {{ label({
-   text: "What is your full name?",
-   isPageHeading: true,
-   size: "l"
+   heading: {
+     id: "custom-id"
+     text: "What is your full name?",
+     level: 2,
+     size: "l"
+   }
  }) }}
```

How's that?

paulrobertlloyd commented:
:shipit:
